File: lib_amap/src/main/java/com/orliu/amap/adapter/RecyclerAdapter.java
package com.orliu.amap.adapter;

import android.content.Context;
import android.support.v7.widget.RecyclerView;
import android.view.ViewGroup;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by Orliu on 2016/10/13.
 */

public class RecyclerAdapter<T> extends RecyclerView.Adapter<ViewHolder> {
    protected Context mContext;
    protected List<T> mDatas;
    protected ItemViewDelegateManager mItemViewDelegateManager;

    public RecyclerAdapter(Context context) {
        mContext = context;
        mItemViewDelegateManager = new ItemViewDelegateManager();
        if (mDatas == null)
            mDatas = new ArrayList<>();
    }

    public void setData(List<T> datas) {
        if (datas != null) {
            mDatas = datas;
            notifyDataSetChanged();
        }
    }

    public void addData(T data){
        mDatas.add(data);
    }

    public void addFirst(T data) {
        if (data != null) {
            mDatas.add(data);
            notifyItemInserted(0);
        }
    }

    public int addLast(T data) {
        if (data != null) {
            mDatas.add(data);
            notifyItemInserted(mDatas.size());
            return getItemCount() - 1;
        }
        return 0;
    }

    public void updateData(int position, T data) {
        notifyItemChanged(position, data);
    }

    public List<T> getDatas() {
        return mDatas;
    }

    @Override
    public int getItemViewType(int position) {
        if (!useItemViewDelegateManager()) return super.getItemViewType(position);
        return mItemViewDelegateManager.getItemViewType(mDatas.get(position), position);
    }


    @Override
    public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        ItemViewDelegate itemViewDelegate = mItemViewDelegateManager.getItemViewDelegate(viewType);
        int layoutId = itemViewDelegate.getItemViewLayoutId();
        ViewHolder holder = ViewHolder.createViewHolder(mContext, parent, layoutId);
        return holder;
    }

    public void convert(ViewHolder holder, T t) {
        mItemViewDelegateManager.convert(holder, t, holder.getAdapterPosition());
    }

    @Override
    public void onBindViewHolder(ViewHolder holder, int position) {
        convert(holder, mDatas.get(position));
    }

    @Override
    public int getItemCount() {
        int itemCount = mDatas.size();
        return itemCount;
    }

    public RecyclerAdapter addItemViewDelegate(ItemViewDelegate<T> itemViewDelegate) {
        mItemViewDelegateManager.addDelegate(itemViewDelegate);
        return this;
    }

    public RecyclerAdapter addItemViewDelegate(int viewType, ItemViewDelegate<T> itemViewDelegate) {
        mItemViewDelegateManager.addDelegate(viewType, itemViewDelegate);
        return this;
    }

    protected boolean useItemViewDelegateManager() {
        return mItemViewDelegateManager.getItemViewDelegateCount() > 0;
    }

}

File: lib_common_kt/src/main/java/com/orliu/kotlin/common/view/rv/RecyclerAdapter.kt
package com.orliu.kotlin.common.view.rv

import android.content.Context
import android.support.v7.widget.RecyclerView
import android.view.ViewGroup
import java.util.*

/**
 * Created by Orliu on 2016/10/13.
 */

open class RecyclerAdapter<T>(private var mContext: Context) : RecyclerView.Adapter<ViewHolder>() {
    private var mDatas: ArrayList<T> = arrayListOf()
    private var mItemViewDelegateManager: ItemViewDelegateManager<T> = ItemViewDelegateManager()

    fun setData(datas: ArrayList<T>?) {
        if (datas != null) {
            mDatas = datas
            notifyDataSetChanged()
        }
        datas?.let {
            mDatas = datas
            notifyDataSetChanged()
        }
    }

    fun update(datas: ArrayList<T>?) {
        datas?.let {
            mDatas = it
            notifyDataSetChanged()
        }
    }

    fun addFirst(data: T?) {
        data?.let {
            mDatas.add(data)
            notifyDataSetChanged()
        }
    }

    fun addLast(data: T?): Int {
        return data?.let {
            mDatas.add(data)
            notifyItemInserted(mDatas.size)
            itemCount - 1
        } ?: let { 0 }
    }

    fun updateItem(position: Int, data: T) {
        notifyItemChanged(position, data)
    }

    override fun getItemViewType(position: Int): Int {
        return if (!useItemViewDelegateManager()) {
            super.getItemViewType(position)
        } else {
            mItemViewDelegateManager.getItemViewType(mDatas[position], position)
        }
    }


    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val itemViewDelegate = mItemViewDelegateManager.getItemViewDelegate(viewType)
        val layoutId = itemViewDelegate.getItemViewLayoutId()
        return ViewHolder.createViewHolder(mContext, parent, layoutId)
    }

    fun convert(holder: ViewHolder, t: T) {
        mItemViewDelegateManager.convert(holder, t, holder.adapterPosition)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        convert(holder, mDatas[position])
    }

    override fun getItemCount(): Int {
        return mDatas.size
    }

    fun addItemViewDelegate(itemViewDelegate: ItemViewDelegate<T>): RecyclerAdapter<*> {
        mItemViewDelegateManager.addDelegate(itemViewDelegate)
        return this
    }

    fun addItemViewDelegate(viewType: Int, itemViewDelegate: ItemViewDelegate<T>): RecyclerAdapter<*> {
        mItemViewDelegateManager.addDelegate(viewType, itemViewDelegate)
        return this
    }

    private fun useItemViewDelegateManager(): Boolean {
        return mItemViewDelegateManager.getItemViewDelegateCount() > 0
    }

}